clean up siac renter download

DavidVorick opened the pull request:
Wasn't happy with the implementation.

I removed the unnecessary 1 second sleep, switched it so that there are no goroutines or channels needed, and I added some extra safety stuff like a timer for a potential infinite loop.

I also added a bunch of TODOs for things that aren't fully done correctly in the code yet.

DavidVorick commented:
I changed the number of tries to a try duration.

I kept it returning an error instead of panicking because in practice that error gets fed directly to a user, and I think it's more friendly to have an error like that than to panic on the user.

DavidVorick commented:
Ok, this code is now to my satisfaction.

ChrisSchinnerl commented on cmd/siac/rentercmd.go:
```diff
@@ -1,5 +1,23 @@
 package main
 
+// TODO: If you run siac from a non-existant directory, the abs() function does
```
I think those are pretty good `programming tasks`. Would be good if you created issues for them too.

ChrisSchinnerl commented on cmd/siac/rentercmd.go:
```diff
@@ -534,48 +553,79 @@ func renterfilesdeletecmd(path string) {
+			continue // benign
+		}
+
+		// Search for the download in the list of downloads.
```
This might also be a good programming task. A new endpoint that let's you fetch a single download.

ChrisSchinnerl commented on cmd/siac/rentercmd.go:
```diff
@@ -534,48 +553,79 @@ func renterfilesdeletecmd(path string) {
+		// give up.
+		if !found {
+			tryTime += OutputRefreshRate
+			if tryTime > time.Minute {
```
Should this also be a constant instead of just `time.Minute` ?

lukechampine replied:
probably clearer to write
```go
start := time.Now()
// ...
    if time.Since(start) > time.Minute {
```

ChrisSchinnerl commented on cmd/siac/rentercmd.go:
```diff
@@ -534,48 +553,79 @@ func renterfilesdeletecmd(path string) {
+		var d api.DownloadInfo
+		found := false
+		for _, d = range queue.Downloads {
+			if d.SiaPath == siapath {
```
I think this is actually not correct. The same `siapath` might be in the download history multiple times.

DavidVorick replied:
It ... should ... grab the most recent one. Otherwise we have no way to distinguish.

DavidVorick replied:
I guess if source and dest are both the same you are okay, we can at least add that

Commit:
clean up siac renter download

Switch to always using async download and then having the progress
function return an error.

Also added some reliability stuff to the download function.

Commit:
switch from a number of tries to a duration of tries

Commit:
repair formatting

Commit:
add constant for timing, clean code up

Also tighten the download to look at destination as well as source.

Commit:
extend a TODO related to constant time download history lookup

Commit:
Merge branch 'master' into siac-dla
